src/backend/gl/src/device.rs

```diff
@@ -468,16 +469,74 @@ pub(crate) unsafe fn set_sampler_info<SetParamFloat, SetParamFloatVec, SetParamI
 impl d::Device<B> for Device {
     unsafe fn allocate_memory(
         &self,
-        _mem_type: c::MemoryTypeId,
+        mem_type: c::MemoryTypeId,
         size: u64,
     ) -> Result<n::Memory, d::AllocationError> {
-        // TODO
-        Ok(n::Memory {
-            properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
-            first_bound_buffer: Cell::new(None),
-            size,
-            emulate_map_allocation: RefCell::new(std::ptr::null_mut()),
-        })
+        if (1 << mem_type.0) & IMAGE_MEM_TYPE_MASK != 0 {
+            assert_ne!((1 << mem_type.0) & DEVICE_LOCAL_MASK, 0);
+            Ok(n::Memory {
+                properties: memory::Properties::DEVICE_LOCAL,
+                buffer: None,
+                target: 0,
+                size,
+                emulate_map_allocation: Cell::new(None),
+                map_flags: 0,
+            })
+        } else {
+            assert!(self.share.private_caps.buffer_role_change);
```

Comment:
wait, so this code can't allocate buffer memory at all if `buffer_role_change` is false?

Reply:
Yeah, I know it's bad, I figured you'd ask about this.  We could either offer a unique type of memory for each role and work differently whether or not `buffer_role_change` was allowed, OR we could force mmap emulation if `buffer_role_change` is not there.  I'm fine with doing either one of those now, this change was just getting to be larger than I expected it to be so I wasn't sure what the best way forward right now was.

Reply:
> OR we could force mmap emulation if buffer_role_change is not there.

Is this really an option though? If the user wants a buffer to support both being a vertex buffer and a uniform buffer, no amount of mmap emulation would give you that.

So it looks to me that we just need to expose this "no role change" policy in the exposed memory types and live with that, for now. Later, if we find that applications are too restricted with this, we could think of some workarounds involving copying data around to temporary buffers, but I don't want us to go there just now.

Reply:
So, just to clarify, you're saying that the gl backend should have memory types based on no role changes at all, or that this should be dynamic depending on `buffer_role_change`?  Doing the first one is... relatively easy, but the second one is a more involved change.  It's easier for me to do the first one and has a higher probability of me getting it right, but if you think the that the gl backend should definitely support multi purpose buffers if available, I can do that.

Reply:
> should be dynamic depending on buffer_role_change

We definitely need the exposed memory types being different, based on the `buffer_role_change` capability.

> is the PIXEL_PACK_BUFFER target special somehow on non-webgl and allowed

We only *have* to use this binding when doing texture-buffer copies. Technically, this is subset of TRANSFER_DST capability (while PIXEL_UNPACK_BUFFER is a subset with TRANSFER_SRC). So with this in mind there is pretty much 1:1 correspondence between a buffer binding point ("role") and `buffer::Usage`.

When the implementation *just* needs to bind a buffer, regardless of the particular bind point, it should pick the bind point that the buffer supports. For example, if the user created a buffer with only VERTEX usage, we'll use the GL_ARRAY_BUFFER binding point for it, etc.